spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/firestore/GcpFirestoreEmulatorAutoConfiguration.java

```diff
@@ -0,0 +1,152 @@
+import java.util.Map;
+
+import com.google.api.gax.grpc.InstantiatingGrpcChannelProvider;
+import com.google.auth.Credentials;
+import com.google.cloud.firestore.FirestoreOptions;
+import com.google.firestore.v1.FirestoreGrpc;
+import io.grpc.ManagedChannel;
+import io.grpc.ManagedChannelBuilder;
+import io.grpc.auth.MoreCallCredentials;
+import reactor.core.publisher.Flux;
+
+import org.springframework.boot.autoconfigure.AutoConfigureBefore;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.cloud.gcp.autoconfigure.firestore.GcpFirestoreAutoConfiguration.FirestoreReactiveAutoConfiguration;
+import org.springframework.cloud.gcp.core.GcpProjectIdProvider;
+import org.springframework.cloud.gcp.data.firestore.FirestoreTemplate;
+import org.springframework.cloud.gcp.data.firestore.mapping.FirestoreClassMapper;
+import org.springframework.cloud.gcp.data.firestore.mapping.FirestoreMappingContext;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+/**
+ * Provides autoconfiguration to use the Firestore emulator if enabled.
+ *
+ * @since 1.2.3
+ * @author Daniel Zou
+ */
+@Configuration
+@ConditionalOnProperty("spring.cloud.gcp.firestore.host-port")
+@AutoConfigureBefore({GcpFirestoreAutoConfiguration.class})
+@EnableConfigurationProperties(GcpFirestoreProperties.class)
+public class GcpFirestoreEmulatorAutoConfiguration {
+
+	private static final String ROOT_PATH_FORMAT = "projects/%s/databases/(default)/documents";
+
+	private final String projectId;
+
+	private final String hostPort;
+
+	private final String firestoreRootPath;
+
+	GcpFirestoreEmulatorAutoConfiguration(
+			GcpFirestoreProperties properties,
+			GcpProjectIdProvider projectIdProvider) throws IOException {
+		this.projectId = projectIdProvider.getProjectId();
+		this.hostPort = properties.getHostPort();
+		this.firestoreRootPath = String.format(ROOT_PATH_FORMAT, this.projectId);
+	}
+
+	@Bean
+	@ConditionalOnMissingBean
+	public FirestoreOptions firestoreOptions() {
+		return FirestoreOptions.newBuilder()
+				.setCredentials(emulatorCredentials())
+				.setProjectId(this.projectId)
+				.setChannelProvider(
+						InstantiatingGrpcChannelProvider.newBuilder()
+								.setEndpoint(this.hostPort)
+								.setChannelConfigurator(input -> input.usePlaintext())
+								.build())
+				.build();
+	}
+
+	private static Credentials emulatorCredentials() {
+		final Map<String, List<String>> headerMap = new HashMap<>();
+		headerMap.put("Authorization", Collections.singletonList("Bearer owner"));
+		headerMap.put("google-cloud-resource-prefix", Collections.singletonList("projects/my-project/databases/(default)"));
```

Comment:
"my-project" -- is it because it's a fake value that's unused? Can you leave a comment about that?

Reply:
Done. I just changed it to plug in the current project-id. I guess this is more technically correct.

Reply:
Or maybe I should make this so the emulator config doesn't require a real project (and not inject a projectIdProvider into the configuration)... Maybe that would make more sense; what do you think?

Reply:
Maybe just hardcode "unused" as project ID?